### PR TITLE
Onboarding Rooms - Join room and create alias

### PR DESCRIPTION
**Pull request by wsarles**
#940

### CODE REVIEW

File: Source/UI/RoomsOnboardingView.swift
Comment (mplorentz) [inline Source/UI/RoomsOnboardingView.swift:125]: Nice! This will be really useful. Can you move it into its own file?

Comment (mplorentz) [reply Source/UI/RoomsOnboardingView.swift:125]: Oh, I think this buttonConfiguration is missing a pressed state.

**Review by mplorentz — COMMENTED** [review]
Wow, this is great work Chad! There are so many moving pieces here, I think you did a good job diving in and stitching them together. I made a list of a few problems I found. Most are small. Going to put them in checkboxes to make them easier to track.
- [ ] The DoneOnboardingStep really leans into the fact that it's the last step 😂 Let's change the "Phew! I'm done" button and the "you made it through in one piece" language. Might need some design guidance from @Chardot here.
- [x] Can we add a dismiss animation when you are finally done with onboarding? Probably `.crossDissolve`.
- [x] I was not able to register "roomtester-555" as an alias on test-room.lorentz.is, but I was able to register "roomtester555". Maybe we are missing something in our client-side validation logic?
- [x] My Planetary wasn't syncing with any peers after going through onboarding. I tracked this down to `AppConfiguration.current` being nil. I didn't see any obvious reason why this would be the case after reading through your code. I think it should be set in the confusingly named function `Onboarding.didStart` but maybe that isn't happening anymore?
- [ ] I'm not sure what the intended design was but I expected to be able to register multiple aliases in multiple rooms, but tapping "Done" after registering one alias dismissed onboarding. I would expect "Done" to take me back to the list of rooms, with maybe another "Done" button on the list of rooms instead of "Skip choosing alias".
- [x] Sometimes an error is displayed for a split second after successful registration of an alias. I see this if I try to register an alias that is already taken, then change my text to a valid alias and press "Done". An error message is displayed for a second before the screen dismisses, but the alias was successfully registered.
- [x] When registering an alias that is already taken I see a long error message for a split second, then it changes to a short error message. Probably we shouldn't show the first one at all? (and ideally we would have a specific error message for when the alias is taken). [video example](https://user-images.githubusercontent.com/1165004/203435013-b84ad84e-cb39-4c0c-b034-66425d9ddcac.mp4) - I am pressing the Done button in this video but you can't tell.
- [x] This PR definitely deserves a CHANGELOG entry!

**Comment by mplorentz** [comment]
> My Planetary wasn't syncing with any peers after going through onboarding. I tracked this down to AppConfiguration.current being nil. I didn't see any obvious reason why this would be the case after reading through your code. I think it should be set in the confusingly named function Onboarding.didStart but maybe that isn't happening anymore?

I'm realizing now that the TestFlight build 387 did something terrible to my keychain. I think this was the real cause of this issue, and not your PR. I'm investigating further but I wouldn't worry about this one for now.

File: Source/Service/RoomInvitationRedeemer.swift
Comment (wsarles) [reply Source/Service/RoomInvitationRedeemer.swift]: Added.

File: Source/UI/Cards/RoomCard.swift
Comment (wsarles) [reply Source/UI/Cards/RoomCard.swift:54]: 😁

File: Source/UI/RoomsOnboardingView.swift
Comment (wsarles) [reply Source/UI/RoomsOnboardingView.swift]: >Nice! This will be really useful. Can you move it into its own file?
Done.

>Oh, I think this buttonConfiguration is missing a pressed state.
Added.

File: Source/UI/Cards/RoomCard.swift
Comment (wsarles) [reply Source/UI/Cards/RoomCard.swift:65]: >Both tapping on a room server and tapping "Back" after choosing a room server have a pretty jarring transition.

Added `withAnimation` on the room selection and back button actions. 
https://github.com/planetary-social/planetary-ios/blob/e960c31059b03faa602efaa179ff36c5202c1fa9/Source/UI/RoomsOnboardingView.swift#L41-L43

File: Source/Model/Room.swift
Comment (wsarles) [reply Source/Model/Room.swift:20]: This allows for a more human readable representation of the room during onboarding. The identifier would be "Lorentz Test Room" in the imported room below:
https://github.com/planetary-social/planetary-ios/blob/e770e62d832bdb5ab83705074c0c2fd65fdadc4f/Source/Configs/Planetary.debug.xcconfig#L64

File: Source/Localization/Localized.swift
Comment (mplorentz) [inline Source/Localization/Localized.swift:305]: When I suggested this copy I was thinking that the purple text would be "Customize your settings" and the black text below it would say "Or just use the defaults." However we are still waiting on feedback/approval from @Chardot here.
Screenshot 2022-11-29 at 4 17 02 PM

File: Source/App/AppController.swift
Comment (mplorentz) [inline Source/App/AppController.swift:33]: This animation felt pretty slow to me. I think Apple's transitions are usually about 0.3 seconds, and most of the ones in Planetary are around that duration as well. Are you ok with matching that?

File: Source/UI/Cards/RoomCard.swift
Comment (mplorentz) [inline Source/UI/Cards/RoomCard.swift:55]: This works, however according to the spec dashes should be valid in an alias. The rooms 2.0 spec points to [RFC 1035](https://www.rfc-editor.org/rfc/rfc1035) which says 
> Note that while upper and lower case letters are allowed in domain
names, no significance is attached to the case.  That is, two names with
the same spelling but different case are to be treated as if identical.

> The labels must follow the rules for ARPANET host names.  They must
start with a letter, end with a letter or digit, and have as interior
characters only letters, digits, and hyphen.  There are also some
restrictions on the length.  Labels must be 63 characters or less.

So maybe the server side is at fault here? This doesn't feel like a blocker to merging this PR, but I think we should open an issue for it if you don't want to spend more time on it here.

**Review by mplorentz — COMMENTED** [review]
Thanks for making all those changes. I had a couple more questions I put below.

Also, according to [this Slack message](https://planetary-app.slack.com/archives/CE6RE7ND6/p1669140182435169) it sounds like we are going to wait to ship this until we have more than 1 room running. If that's the case I think we'll want to wait to merge this PR until then so that `main` stays deployable :( Although imo it seems better to ship this with just 1 room.

File: Source/Localization/Localized.swift
Comment (wsarles) [reply Source/Localization/Localized.swift]: Updated the copy with your suggestion for now.

File: Source/UI/Cards/RoomCard.swift
Comment (wsarles) [reply Source/UI/Cards/RoomCard.swift:55]: >So maybe the server side is at fault here? 

Submitting an alias with hyphens anywhere in the string would cause a server side error.

Comment (wsarles) [reply Source/UI/Cards/RoomCard.swift:55]: Additionally, I was allowed to create an alias starting with a digit.

image

File: Source/App/AppController.swift
Comment (wsarles) [reply Source/App/AppController.swift]: Fixed.

**Comment by rabble** [comment]
Waiting on finalizing the default room options and then running through testing in testflight.

**Comment by mplorentz** [comment]
@wsarles rather than leaving this PR open forever, what if modify it so that it just shows test-room.lorentz.is if you create an account on the test network. Then we could merge this one and open a new one to show the screens for the main network when we have the servers in order.

**Comment by wsarles** [comment]
>  what if modify it so that it just shows test-room.lorentz.is if you create an account on the test network...

Sounds good.

**Review by mplorentz — APPROVED** [review]
Just one more tweak before we finally merge this! Let's change the `primaryAction` button text on the DoneOnboardingStep back to what it was before since in production it'll still be the last step for now. I'm going to go ahead and click approve though since it's just a minor copy change. Feel free to click merge once that is done as long as CI is happy. Thanks for being so patient on this!

**Pull request merged**
